controllers/shooter.py

```diff
@@ -75,54 +69,54 @@ def searching(self) -> None:
         if self.vision.get_vision_data()[2] is not None:
             # means no data is available
             # print(f"searching -> tracking {self.vision.get_vision_data()}")
-            # self.next_state("tracking")
-            self.state = self.tracking
+            self.next_state("tracking")
 
-    # @state
-    def tracking(self) -> None:
+    @state
+    def tracking(self, initial_call) -> None:
         """
         Aiming towards a vision target and spining up flywheels
         """
+        if initial_call:
+            self.shooter.stop_motors()
         dist, delta_angle, timestamp = self.vision.get_vision_data()
         # collect data only once per loop
         if timestamp is None:
-            # self.next_state("searching")
+            self.next_state("searching")
             # print(f"tracking -> searching {self.vision.get_vision_data()}")
-            self.state = self.searching
         else:
             if abs(delta_angle) > self.find_allowable_angle(dist):
                 # print(f"Telling turret to slew by {delta_angle}")
                 self.turret.slew(delta_angle)
             if self.ready_to_spin():
-                # self.next_state("firing")
-                # print(f"tracking -> spining_up {self.vision.get_vision_data()}")
                 self.distance = dist
-                self.state = self.spining_up
+                self.next_state("spining_up")
+                # print(f"tracking -> spining_up {self.vision.get_vision_data()}")
 
-    def spining_up(self) -> None:
-        if self.initial_call:
+    @state
+    def spining_up(self, initial_call) -> None:
+        if initial_call:
             self.shooter.set_range(self.distance)
-            self.initial_call = False
-        if self.ready_to_fire() and self.input_command:
-            self.distance = None
-            self.initial_call = True
-            # print(f"spining_up -> firing {self.vision.get_vision_data()}")
-            self.state = self.firing
-
-    # @state
+        if self.turret.is_ready():
+            if self.ready_to_fire() and self.fire_command:
+                self.distance = None
+                # print(f"spining_up -> firing {self.vision.get_vision_data()}")
+                self.next_state("firing")
+        else:
+            self.next_state("tracking")
+
+    @state
     def firing(self) -> None:
         """
         Positioned to fire, inject and expel a single ball
         """
         self.shooter.fire()
-        # self.next_state("tracking")
-        self.state = self.tracking
+        self.next_state("spining_up")
```

Comment:
Ok, we go from firing to spinning_up so we can keep shooting as long as we have balls to shoot. But the only way to leave this two-state loop seems to be if the turret stops being ready. But the turret will remain ready as long as it's completed the last slew command, and we can't give it a new one until we stop spinning, so I don't see how this isn't an infinite loop. Is the MagicBot state machine doing some magic I'm missing?

Also, I think we need a state for when we're looking for balls to intake. The robot is moving so we can't really do searching or tracking, and we aren't firing or spinning up the shooter, either.